### ISSUE TITLE
numpy.longdouble (emscripten `long double`) not actually handled by most math library functions

### ISSUE THREAD

**Issue by WarrenWeckesser** [pyodide/pyodide#4881]
## 🐛 Bug

You might be aware of this, but I only recently started experimenting with emscripten and pyodide, and I ran into an issue with the `long double` type.  In emscripten, `long double` is IEEE float128, but its math library is based on musl, and musl does not implement float128 algorithms for many of its functions.  It looks like the C functions `sinl`, `cosl` and `tanl` handle float128, but for `logl`, `log1pl`, `expl`, `expm1l`, `powl` and others, if it detects float128 (i.e. `LDBL_MANT_DIG == 113`), it downcasts and uses its `double` implementation (silently, with no warning).

For example, `np.exp(np.longdouble(1))` should be `2.71828182845904523536028747135266`, but in pyodide, we get

```
>>> np.exp(np.longdouble(1))
2.718281828459045090795598298427649
```

The following shows that the above is equivalent to doing the calculation with double precision instead of float128:

```
>>> np.longdouble(np.exp(1))
2.718281828459045090795598298427649
```

This is a surprising "gotcha" for the folks who try to use `np.longdouble` in pyodide.

### Environment

- Pyodide Version: whatever is currently running at https://pyodide.org/en/stable/console.html
- Browser version: Firefox 126.0

**Comment by hoodmane** [comment]
Thanks for the report. cc @agriyakhetarpal @rgommers. Maybe if someone has time we can work with upstream to add 128 float support. One question: does numpy distribute musl Linux wheels? Do they have the same problem? If so how do we deal with that.

**Comment by WarrenWeckesser** [comment]
I haven't been involved with the wheel building and distribution infrastructure for numpy, but at https://pypi.org/project/numpy/#files I see several musl-based wheels.

The problem reported here only occurs if the implementation of `long double` in the compiler is IEEE float128 (as it is with emscripten). It looks like the functions in musl can handle the more common case where `long double` is the same as `double` and where `long double` is 80 bit extended precision. For example, here is musl's code for `expl`: https://git.musl-libc.org/cgit/musl/tree/src/math/expl.c.  You can see that it decides which implementation to use based on the C macro `LDBL_MANT_DIG`.  It implements appropriate algorithms for `LDBL_MANT_DIG == 53` (where `long double` is the same as `double`), and `LDBL_MANT_DIG == 64` (80 bit extended precision).  With emscripten, `LDBL_MANT_DIG` is 113, and down at the bottom of that file is

```
#elif LDBL_MANT_DIG == 113 && LDBL_MAX_EXP == 16384
// TODO: broken implementation to make things compile
long double expl(long double x)
{
	return exp(x);
}
#endif
```

In other words, with IEEE float128, fall back to the double precision version.
